backend/src/main/java/gov/cdc/usds/simplereport/service/OktaService.java

```diff
@@ -0,0 +1,121 @@
+package gov.cdc.usds.simplereport.service;
+
+import org.springframework.stereotype.Service;
+
+import java.util.Map; 
+import java.util.HashMap;
+
+import org.springframework.beans.factory.annotation.Value;
+
+import com.okta.sdk.client.Client;
+import com.okta.sdk.client.Clients;
+import com.okta.sdk.resource.user.User;
+import com.okta.sdk.resource.user.UserBuilder;
+import com.okta.sdk.resource.group.GroupList;
+import com.okta.sdk.resource.group.GroupType;
+import com.okta.sdk.resource.group.GroupBuilder;
+import com.okta.sdk.authc.credentials.TokenClientCredentials;
+
+import gov.cdc.usds.simplereport.db.model.ApiUser;
+
+import gov.cdc.usds.simplereport.config.authorization.OrganizationRole;
+
+/**
+ * Handles all user/organization management in Okta
+ */
+@Service
+public class OktaService {
+
+    @Value("${okta.entity-mgmt.org-url}")
+    private String ORG_URL;
+    @Value("${OKTA_API_KEY}")
+    private String API_TOKEN;
+    @Value("${simple-report.authorization.role-prefix}")
+    private String ROLE_PREFIX;
+
+    Client _client;
+
+
+    public OktaService() {
+        _client = Clients.builder()
+            .setOrgUrl(ORG_URL) 
+            .setClientCredentials(new TokenClientCredentials(API_TOKEN))
+            .build();
+    }
+
+    public void createUser(ApiUser apiUser) {
+        Map<String,Object> userProfileMap = new HashMap<String,Object>();
+        userProfileMap.put("firstName",apiUser.getNameInfo().getFirstName());
+        userProfileMap.put("middleName",apiUser.getNameInfo().getMiddleName());
+        userProfileMap.put("lastName",apiUser.getNameInfo().getLastName());
+        // This is cheating. Suffix and honorific suffix aren't the same thing. Shhh.
+        userProfileMap.put("honorificSuffix",apiUser.getNameInfo().getSuffix());
+        // We assume login == email
+        userProfileMap.put("email",apiUser.getLoginEmail());
+        userProfileMap.put("login",apiUser.getLoginEmail());
+
+        UserBuilder.instance()
+                .setProfileProperties(userProfileMap)
+                .buildAndCreate(_client);
+    }
+
+    public void updateUser(String oldUsername, ApiUser apiUser) {
```

Comment:
as far as i'm concerned, any time we update a User record in postgres, we should be updating the corresponding attributes of that user in Okta, assuming Okta knows those attributes at all. Most important for email, related to our discussion above. Is there a downside in having this method?

Reply:
I agree that things should be in sync! (Right now they aren't necessarily synced, as far as that goes—probably worth fixing but not urgent.) I think the question is how changes can be made: the original assumption was that Okta was the system of record, and all changes were made through Okta (or through a federated IDP), and then mirrored to SimpleReport by way of the authentication token. This gives SimpleReport the responsibility for managing the Okta user details, rather than being a passive receiver of them—I don't know if that is a thing we want to do or not, setting aside for the moment the email-as-durable-ID problem.

Reply:
if changes are made only through Okta, is there any way we can keep these synced? i feel like if we can't update Okta when SimpleReport users update their data, then we shouldn't offer an updateUser mutation at all. thoughts?